### PR TITLE
Vschema DDL: add column, drop column, show columns etc.

### LINKED ISSUE #4641: enhancement of vschema alter to better support authoritative columns list
This is a propose to further enhancement of: https://github.com/vitessio/vitess/issues/3459

#### Feature Description

We'd like to have more `ALTER VSCHEMA ON {table}` syntax to:
1. set a table with `authoritative_columns_list` enabled in vschema
2. add/drop a column from a table which is `authoritative_columns_list` enabled in vschema
The motivation is to allow application to mix vschema ddl as above with common sql ddls that either creates new table, or add/remove columns from a table w/o doing a whole vchema swap, while the `authoritative_columns_list` feature is on.
At Pinterest, we're using the `authoritative_columns_list` feature, and the most common DDL that impacts the vschema tables are `CREATE TABLE`, `ALTER TABLE ADD COLUMN`. 

#### Use Case(s)

1. after a `CREATE TABLE` ddl, we could enable authoritative columns list feature by:
* `ALTER VSCHEMA ON {table} ADD COLUMN c1 {type}`
* `ALTER VSCHEMA ON {table} ADD COLUMN c2 {type}`
* ...
* `ALTER VSCHEMA SET TABLE {table} AUTHORITATIVE`
2. after an `ALTER TABLE {table} ADD c1 {type}`
* `ALTER VSCHEMA ON {table} ADD COLUMN c1 {type}`
3. after an `ALTER TABLE {table} DROP c1`
* `ALTER VSCHEMA ON {table} DROP COLUMN c1`

given above support, an application schema change with `authoritative_columns_list` won't have to go through an entire vschema swap, just need to use the additional vschema ddl. that could also facilitate the DBA review process, as they don't have to review one *.sql, and *.json separately. lastly, there's a trickiness with the columns change with `authoritative_columns_list`:
1. when adding a column, we want the vschema column add to happen after the table schema change
2. when dropping a column, we want the vschema column to be droppped before the table schema change
the sequence is very difficult to be enforced if running *.sql & *.json change in separation, but as the vschema ddl could get mixed into the *.sql ddl, the sequence of the statement could reflect the above accordingly

### PR DESCRIPTION
**Pull request by inexplicable**
`ALTER VSCHEMA ON {table} ADD COLUMN {col} {type}`
`ALTER VSCHEMA ON {table} DROP COLUMN {col}`
`ALTER VSCHEMA ON {table} SET authoritative={TRUE|FALSE}`
`ALTER VSCHEMA ON {table} SET column_list_authoritative={TRUE|FALSE}`
`SHOW VSCHEMA COLUMNS ON {table}`

Fixes #4641

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -741,22 +741,31 @@ type DDL struct {
...
+	RenameStr            = "rename"
+	TruncateStr          = "truncate"
+	FlushStr             = "flush"
+	CreateVindexStr      = "create vindex"
+	DropVindexStr        = "drop vindex"
+	AddVschemaTableStr   = "add vschema table"
+	DropVschemaTableStr  = "drop vschema table"
+	AddColVindexStr      = "on table add vindex"
+	DropColVindexStr     = "on table drop vindex"
+	AddVschemaColStr     = "on table add column"
```
Comment (inexplicable) [inline go/vt/sqlparser/ast.go]: > per dweitzman
> Nit: columns in a vindex is not necessarily authoritative. Depends whether the table is marked as having authoritative columns. Maybe something like "AddVschemaColStr" would be a more accurate name than "AddAuthColumnStr"

**Comment by inexplicable** [comment]
@sougou i started off using `authoritative`, then @dweitzman suggested it to be aligned with the struct field naming `column_list_authoritative`, i guess i found that too wordy, and wanna keep a shorthand value. if we're going to land on `is_authoritative`, that should be fine, will just wait for this to be confirmed. and i'll make it a single option.

**Comment by sougou** [comment]
In that case, we should definitely not overload `is_authoritative` because the name is reserved for future use.

File: go/vt/sqlparser/sql.y [hunk 2]
```diff
@@ -1338,6 +1338,30 @@ alter_statement:
         },
       }
   }
+| ALTER VSCHEMA ON table_name ADD COLUMN column_definition
```
[the remark this answers is not in the document]
Comment (demmer) [reply go/vt/sqlparser/sql.y]: Since the only element we care about from column_definition is the type, I’d lean towards just parsing that and not the rest of the syntax.

File: go/vt/topotools/vschema_ddl.go [hunk 3]
```diff
@@ -180,6 +180,59 @@ func ApplyVSchemaDDL(ksName string, ks *vschemapb.Keyspace, ddl *sqlparser.DDL)
...
+		ks.Tables[tableName] = table
+		return ks, nil
+
+	case sqlparser.SetVschemaUpdatesStr:
+		name := ddl.Table.Name.String()
+		if _, ok := ks.Tables[name]; !ok {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vschema does not contain table %s in keyspace %s", name, ksName)
+		}
+		for _, update := range ddl.VschemaUpdates {
+			// for now the only setting we accept is `authoritative` or `column_list_authoritative`
```
[the remark this answers is not in the document]
Comment (demmer) [reply go/vt/topotools/vschema_ddl.go]: I agree and don’t think we should support two types for the same setting.

Being verbose with column_list_authoritative is clearest.

File: go/vt/topotools/vschema_ddl.go [hunk 4]
```diff
@@ -180,6 +180,59 @@ func ApplyVSchemaDDL(ksName string, ks *vschemapb.Keyspace, ddl *sqlparser.DDL)
...
+			if colName := update.Name.Name; colName.EqualString("authoritative") || colName.EqualString("column_list_authoritative") {
```
Comment (demmer) [inline go/vt/topotools/vschema_ddl.go]: Minor nit — this isn’t a column name so something like `setting` would be clearer than `colName`

File: go/vt/vtgate/executor_test.go [hunk 5]
```diff
@@ -1254,6 +1254,51 @@ func waitForColVindexes(t *testing.T, ks, table string, names []string, executor
 	return nil
 }
 
+func waitForVschemaCol(t *testing.T, ks, table string, name string, drop bool, executor *Executor) (*vschemapb.SrvVSchema, *vschemapb.Column) {
```
Comment (dweitzman) [inline go/vt/vtgate/executor_test.go]: This helper is modeled after some existing helpers, and these similar-looking helps look like they're multiplying. Could be cleaner to replace all the similar-ish wait logic with a single, more generic wait helper (not necessarily in this PR, though). Maybe something like this:

```
vschemaWaiter := makeVschemaWaiter(t, executor.serv, "aa")
_, _ := executor.Execute(... some vschema change ...)
vschema := vschemaWaiter()
// validate that the vschema looks as it should
// run another execute that changes the vschema
vschema = vschemaWaiter()
// validate again
// etc.
```

Where makeVschemaWaiter() might look something like:

```
// makeVschemaWaiter returns a function that gets the current vschema.
func makeVschemaWaiter(t testing.T, s svtopo.Server, cell string) func() *vschemapb.SrvVSchema {
    t.Helper()
	c := make(chan *vschemapb.SrvVSchema, 20)
    sentinel := &vschemapb.SrvVSchema{}
	s.WatchSrvVSchema(context.Background(), cell, func(v *vschemapb.SrvVSchema, err error) {
		c <- v
	})
    return func() *vschemapb.SrvVSchema {
       c <- sentinel
       var lastFound *vschemapb.SrvVSchema
       for {
          select {
            case found := <-c:
                if found == sentinel {
                  return lastFound
                }
                lastFound = found
            case <-time.After(100 * time.Millisecond):
                t.Error("Timeout waiting for vschema update")
                return nil
            }
       }
    }
}
```

File: go/vt/vtgate/executor_test.go [hunk 6]
```diff
@@ -1718,6 +1763,146 @@ func TestExecutorAddDropVindexDDL(t *testing.T) {
...
+	_, err := executor.Execute(context.Background(), "TestExecute", session, stmt, nil)
+	if err != nil {
+		t.Error(err)
+	}
+	vschema = waitForVschemaTables(t, ks, append(vschemaTables, "test_table"), executor)
+
+	if vschema.Keyspaces[ks].Tables["test_table"].ColumnListAuthoritative {
+		t.Fatalf("test_table should not be authoritative yet")
+	}
+	stmt = "alter vschema on test_table set column_list_authoritative=true"
```
[the remark this answers is not in the document]
Comment (inexplicable) [reply go/vt/vtgate/executor_test.go]: 👍

File: go/vt/vtgate/executor_test.go (hunk 5 above)
Comment (inexplicable) [reply go/vt/vtgate/executor_test.go]: make sense to me

Comment (inexplicable) [reply go/vt/vtgate/executor_test.go]: hmm, 2nd thought, as the vschema ddl stuff is to be replaced, i don't want to spend too much renovation on the test case.

File: go/vt/vtgate/executor_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (inexplicable) [reply go/vt/vtgate/executor_test.go]: done

File: go/vt/vtgate/executor.go [hunk 7]
```diff
@@ -940,6 +943,46 @@ func (e *Executor) handleShow(ctx context.Context, safeSession *SafeSession, sql
...
+		}
+
+		for _, col := range table.Columns {
+			rows = append(rows, buildVarCharRow(col.Name, col.Type.String()))
+		}
+
+		return &sqltypes.Result{
+			Fields:       buildVarCharFields("Name", "Type"),
+			Rows:         rows,
+			RowsAffected: uint64(len(rows)),
```
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vtgate/executor.go]: The current behavior mimics the behavior of the mysql clients, which has two modes: `use_results` and `store_results`. One is streaming and the other is all-at-once. The all-at-once case does set `RowsAffected` even for selects, but the all-at-once case doesn't. Since this is a client side behavior (which doesn't parse queries), it's likely that it doesn't differentiate between select, show, etc.

Right now, I think this is relevant only for grpc. There's probably no way to send this info when returning query results.

File: go/vt/vtgate/executor_test.go [hunk 8]
```diff
@@ -1718,6 +1763,158 @@ func TestExecutorAddDropVindexDDL(t *testing.T) {
...
+		t.Error(err)
+	}
+	vschema, _ = waitForVschemaUpdates(t, ks, "test_table", false, executor)
+	if vschema.Keyspaces[ks].Tables["test_table"].ColumnListAuthoritative {
+		t.Fatalf("test_table should not be authoritative now")
+	}
+
+	stmt = "alter vschema on test_table set column_list_authoritative=invalid"
+	_, err = executor.Execute(context.Background(), "TestExecute", session, stmt, nil)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (inexplicable) [reply go/vt/vtgate/executor_test.go]: 👍

[the remark this answers is not in the document]
Comment (inexplicable) [reply go/vt/vtgate/executor_test.go]: done & squashed into one commit

**Comment by sougou** [comment]
Is this ready? If so, @dweitzman: can you do a final once-over and merge?

**Comment by inexplicable** [comment]
> Is this ready? If so, @dweitzman: can you do a final once-over and merge?

yes, i believe the comments were all addressed (or reject 🤣 )

**Comment by sougou** [comment]
@dweitzman: waiting on your final approval.

**Comment by dkhenry** [comment]
May I request we update the title of this PR before its merged. Right now its not very descriptive

**Comment by sougou** [comment]
@inexplicable conflicts need merging, and @dkhenry requests a better title.

**Comment by morgo** [comment]
@inexplicable followup ping :-)
